networkx/algorithms/bipartite/envy_free_matching.py

```diff
@@ -0,0 +1,375 @@
+
+    top_nodes: list
+      if graph is not all connected top_nodes is the set of the top nodes in
+      the bipartite graph `G`.
+
+    Returns
+    -------
+    EFM: tuple of sets
+      The partition returns as a tuple of 4 sets of nodes:
+      `(X_L, X_S, Y_L, Y_S)` where X_L and Y_L are the "good nodes" of G
+      while X_S and Y_S are the "bad nodes" of G where no envy-free matching
+      exists according to THM 1.3 in the article.
+
+    Examples
+    --------
+    Example 1: Perfect matching
+
+    >>> G = nx.complete_bipartite_graph(3, 3)
+    >>> M = nx.bipartite.hopcroft_karp_matching(G)
+    >>> nx.bipartite.envy_free_matching_partition(G, M=M)
+    ({0, 1, 2}, set(), {3, 4, 5}, set())
+
+    Where there exists a perfect matching the maximum cardinality
+    envy free matching is the perfect matching.
+
+    Example 2: Non-empty envy-free matching
+
+    >>> G = nx.Graph([(0, 3), (0, 4), (1, 4), (2, 4)])
+    >>> M = {0: 3, 3: 0, 1: 4, 4: 1}
+    >>> nx.bipartite.envy_free_matching_partition(G, M=M)
+    ({0}, {1, 2}, {3}, {4})
+
+    Here the graph contains a non-empty envy-free matching so X_L and Y_L are not empty.
+
+    Example 3: Odd path
+
+    >>> G = nx.Graph([(0, 3), (1, 3), (1, 4), (2, 4)])
+    >>> M = {0: 3, 3: 0, 4: 1, 1: 4}
+    >>> nx.bipartite.envy_free_matching_partition(G, M=M)
+    (set(), {0, 1, 2}, set(), {3, 4})
+
+    Like presented in the article, odd path contains an empty envy-free matching
+    so X_L and Y_L are empty in the partition.
+
+    Example 4: Y-path-saturated graph
+
+    >>> G = nx.Graph(
+    ...     [(0, 6), (1, 6), (1, 7), (2, 6), (2, 8), (3, 9), (3, 6), (4, 8), (4, 7), (5, 9)]
+    ... )
+    >>> M = {0: 6, 6: 0, 1: 7, 7: 1, 2: 8, 8: 2, 3: 9, 9: 3}
+    >>> nx.bipartite.envy_free_matching_partition(G, M=M)
+    (set(), {0, 1, 2, 3, 4, 5}, set(), {8, 9, 6, 7})
+
+    Like presented in the article [1], Y-path-saturated graph contains an empty
+    envy-free matching so X_L and Y_L are empty in the partition.
+
+    References
+    ----------
+    .. [1] "Envy-free Matchings in Bipartite Graphs and their Applications
+    to Fair Division", by Elad Aigner-Horev and Erel Segal-Halevi (2022),
+    https://arxiv.org/abs/1901.09527
+    Algorithm 1: Finding the EFM partition of a bipartite graph.
+    """
+    if M is None:
+        M = maximum_matching(G, top_nodes=top_nodes)
+    X, Y = nx.bipartite.sets(G, top_nodes=top_nodes)
+    X_subsets, Y_subsets = _M_alternating_sequence(G, M, top_nodes=top_nodes)
+    X_S = set(it.chain.from_iterable(X_subsets))
+    Y_S = set(it.chain.from_iterable(Y_subsets))
+    return set(X) - X_S, X_S, set(Y) - Y_S, Y_S
+
+
+@nx.utils.not_implemented_for("directed")
+@nx.utils.not_implemented_for("multigraph")
+def maximum_envy_free_matching(G, *, top_nodes=None):
+    r"""Return an envy-free matching of maximum cardinality of the bipartite graph `G`
+
+    A matching in a bipartite graph with bipartite sets X and Y is called envy-free,
```

Comment:
Since we already defined EFM at the top of the file, I am not sure it is needed to redefine it here (but it may still be good to redefine, for people who read only this specific function).